add start_index to dataloader

janewangfb opened the pull request:
Summary: this is prepartion for EDPM. As new nodes joins/leaves, we will need to read data from some middlepoint of the dataset.

Differential Revision: D13628096

soumith commented:
cc: @apaszke @colesbury @ssnl for review.

Looks reasonable to me in theory.

Also @janewangfb lint is broken, see the test failure.

SsnL commented:
This is reasonable to support. But I’m not so sure about adding the kwarg to DataLoader. Shouldn’t a custom sampler be sufficient?

apaszke reviewed (changes requested):
This can already be implemented using samplers, or by passing in a dataset that shifts indices by the reuiqred amount and calls into another dataset. We shouldn’t expand the API unnecessarily, especially that it’s already super complicated.

janewangfb commented:
@soumith fixed the lint issue.

janewangfb commented:
@SsnL DataLoader does not necessary need to specify a sampler. when it is not specified during initialization, DataLoader __init__ will create a new sampler, in this case, we will need to pass the start_index in.

janewangfb commented:
@apaszke This implementation is actually adding a start_index to sampler. But for dataloader, it could create a sampler in its __init__ method. in this case, we will need to pass in the start_index through dataloader.

soumith commented:
@janewangfb I think what Adam is advocating, which does make sense is to have an `EDPMSampler` that would do the shifting, and pass that in to the DataLoader.

I think this would be needed to be done as we need more features for complex training.
I'm thinking -- it might be the case that we need to record a cursor into the dataset which itself does not have a finite length, etc.

soumith commented:
so, for example:

```
sampler = EDPMSampler(RandomSampler(...), start_index=35)
loader = DataLoader(...., sampler=sampler)
```

janewangfb commented:
@soumith, thanks for the clarification. I updated this PR to implement EDPMSampler.

soumith commented:
@janewangfb thanks. EDPMSampler does not need to live in core at this stage, as none of the other elastic pieces have come in.
Can you put it out of core, and maybe in the elastic DP package (if there's one) for now, and we'll pull it into core with proper naming and API afterwards?

janewangfb committed:
add EDPMSampler (#15935)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/15935

this is prepartion for EDPM. As new nodes joins/leaves, we will need to read data from some middlepoint of the dataset.

Differential Revision: D13628096

fbshipit-source-id: 9f53ea5a31127cc340872642ab0a90864f127739

janewangfb commented:
@soumith I moved the code to torch/fb/elastic. But it seems only one FILE is exported to the PR. Can you review D13628096 instead?
